File: cpp/839.similar-string-groups.cpp
class DSU {
private:
    vector<int> parent;
    vector<int> rank;

public:
    DSU(int size): rank(vector<int>(size, 1)), parent(vector<int>(size, 0)) {
        iota(parent.begin(), parent.end(), 0);
    }

    int find(int x) {
        while (x != parent[x]) {
            parent[x] = parent[parent[x]];
            x = parent[x];
        }
        return x;
    }

    bool join(int x, int y) {
        auto par_x = find(x);
        auto par_y = find(y);

        if (par_x == par_y) {
            return false;
        }

        if (rank[par_y] > rank[par_x]) {
            swap(par_x, par_y);
        }

        parent[par_y] = parent[par_x];
        rank[par_x] += rank[par_y];
        rank[par_y] = 0;
        return true; 
    }

    int components() {
        int res = 0;
        for (int i = 0; i < parent.size(); i++) {
            if (find(i) == i) {
                res++;
            } 
        }
        return res;
    }
};

class Solution {
private:
    bool checkSimilar(string& s1, string& s2) {
        if (s1 == s2) {
            return true;
        }
        int replacements = 0;

        for (int i = 0; i < min(s1.size(), s2.size()); i++) {
            if (s1[i] != s2[i]) {
                replacements++;
            }

            if (replacements >= 3) {
                return false;
            } 
        }
        return true;
    }

public:
    int numSimilarGroups(vector<string>& strs) {
        int n = strs.size();
        DSU dsu = DSU(n);

        for (int i = 0; i < n; i++) {
            for (int j = i + 1; j < n; j++) {
                if (checkSimilar(strs[i], strs[j])) {
                    dsu.join(i, j);
                }
            }
        }
        return dsu.components();
    }
};

File: golang/839.similar-string-groups.go
func numSimilarGroups(strs []string) int {
    n := len(strs)
    dsu := NewDSU(n)

    for i := 0; i < n; i++ {
        for j := i + 1; j < n; j++ {
            if similar(strs[i], strs[j]) {
                _ = dsu.join(i, j)
            }
        }
    }

    return dsu.count()
}

func similar(s1, s2 string) bool {
    if s1 == s2 {
        return true
    }
    replacements := 0
    n := min(len(s1), len(s2))
    for i := 0; i < n; i++ {
        if s1[i] != s2[i] {
            replacements++
        }

        if replacements >= 3 {
            return false
        }
    }
    return true
}

func min(a, b int) int {
    if a < b {
        return a
    }
    return b
}

type DSU struct {
    parent []int
    rank []int
}

func NewDSU(size int) *DSU {
    rank := make([]int, size)
    parent := make([]int, size)
    for i := range rank {
        parent[i] = i
        rank[i] = 1
    }

    return &DSU{ parent, rank }
}

func (d *DSU) find(x int) int {
    for x != d.parent[x] {
        d.parent[x] = d.parent[d.parent[x]]
        x = d.parent[x]
    }
    return x
}

func (d *DSU) join(x, y int) bool {
    par_x, par_y := d.find(x), d.find(y)

    if par_x == par_y {
        return false
    }

    if d.rank[par_x] < d.rank[par_y] {
        par_x, par_y = par_y, par_x
    }

    d.parent[par_y] = d.parent[par_x]
    d.rank[par_x] += d.parent[par_y]
    d.rank[par_y] = 0
    return true
}

func (d *DSU) count() int {
    count := 0
    for i := range d.parent {
        if d.find(i) == i {
            count++
        }
    }
    return count
}